doc/sr_log2source.8.rst

==============
 SR_Log2source
==============

-------------------------------------------------------
Copy Log Messages from xlog to the Corresponding Source 
-------------------------------------------------------

:Manual section: 1 
:Date: @Date@
:Version: @Version@
:Manual group: MetPx Sarracenia Suite


SYNOPSIS
========

**sr_report2source** -b <broker> -i <instances> start|stop|restart|reload|status

DESCRIPTION
===========

The administrator daemon to copy log messages to the relevant source.
**sr_report2source** subscribes to all log messages 
(posted on the xlog exchange), ignoring the ones not from the current cluster, 
and copying those that concern "source" to the xs\_"source"' exchange, where
the user concerned can read them.

The log notification protocol is defined here `sr_report(7) <sr_report.7.html>`_

**sr_report2source** connects to a *broker* and subscribes to all log messages. 
It uses the **from_cluster** 
field in the notification header to make sure it will process its logs. It uses
the **source** field in the notification header to post to exchange 'xl\_'source
the log messages. Any other log messages are skipped.

The **sr_report2source** command can takes 3 arguments: a broker, a number of instances,
followed by an action start|stop|restart|reload|status... (self described).

CONFIGURATION
=============

In general, the options for this component are described by the
`sr_config(7) <sr_config.7.html>`_  page which should be read first. 
It fully explains the option configuration language, and how to find 
the option settings.

The broker option sets all the credential information to connect to 
the **AMQP** server: 

**--broker|-b amqp{s}://<user>:<pw>@<brokerhost>[:port]/<vhost>**

::

      (default: amqp://guest:guest@localhost/ ) 

Once connected to an AMQP broker, **sr_report2source** it will process all log messages. 




SEE ALSO
========

`sr_config(7) <sr_config.7.html>`_ - the format of configurations for MetPX-Sarracenia.

`sr_report(7) <sr_report.7.html>`_ - the format of log messages.

`sr_post(1) <sr_post.1.html>`_ - post announcemensts of specific files.

`sr_post(7) <sr_post.7.html>`_ - The format of announcements.

`sr_subscribe(1) <sr_subscribe.1.html>`_ - the http-only download client.

`sr_watch(1) <sr_watch.1.html>`_ - the directory watching daemon.

`sr_report2clusters(8) <sr_report2clusters.8.html>`_ - the inter-cluster log copy daemon.



sarra/sr_log2source.py
#!/usr/bin/python3
#
# This file is part of sarracenia.
# The sarracenia suite is Free and is proudly provided by the Government of Canada
# Copyright (C) Her Majesty The Queen in Right of Canada, Environment Canada, 2008-2015
#
# Questions or bugs report: dps-client@ec.gc.ca
# sarracenia repository: git://git.code.sf.net/p/metpx/git
# Documentation: http://metpx.sourceforge.net/#SarraDocumentation
#
# sr_log2source.py : python3 program takes all log messages and repost them to the
#                    log exchange of the source user
#
#
# Code contributed by:
#  Michel Grenier - Shared Services Canada
#  Last Changed   : Dec 17 09:20:42 EST 2015
#
########################################################################
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful, 
#  but WITHOUT ANY WARRANTY; without even the implied warranty of 
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
#
#

#============================================================
# usage example
#
# sr_log2source [options] [config] [start|stop|restart|reload|status]
#
# logs message have travel back to this cluster
#
# conditions :
# exchange                = xreport
# topic                   = v02.report
# header['from_cluster']  = cluster        (here)
# header['source]         in source_users  (one of our source : users.conf)
#
# it is a log message that our source sould be able to see so:
#
# publish this log message in xr_"source"
#
#============================================================

try :    
         from sr_consumer       import *
         from sr_instances      import *
         from sr_message        import *
except : 
         from sarra.sr_consumer  import *
         from sarra.sr_instances import *
         from sarra.sr_message   import *

class sr_log2source(sr_instances):

    def __init__(self,config=None,args=None):
        sr_instances.__init__(self,config,args)

    def check(self):

        # create bindings from default ?

        if self.bindings == [] :
           key = self.topic_prefix + '.' + self.subtopic
           self.bindings.append( (self.exchange,key) )

        # scan users for role source

        self.source_users = []

        for user in self.users :
            roles = self.users[user]
            if 'source' in roles :
               self.source_users.append(user)

        self.logger.debug("source_users = %s " % self.source_users)

        if len(self.source_users) == 0 :
           self.nbr_instances = 0

    def close(self):
        self.consumer.close()

    def connect(self):

        # =============
        # create message if needed
        # =============

        self.msg = sr_message(self)

        # =============
        # consumer  queue_name : let consumer takes care of it
        # =============

        self.consumer = sr_consumer(self)

        # =============
        # publisher... (publish back to consumer)  
        # =============

        self.publisher = self.consumer.publish_back()

        # =============
        # setup message publisher
        # =============

        self.msg.publisher = self.consumer.publisher
        self.msg.post_exchange_split = self.post_exchange_split

    def help(self):
        self.logger.info("Usage: %s [options] [config] [start|stop|restart|reload|status]  \n" % self.program_name )


    # =============
    # __on_message__  internal message validation
    # =============

    def __on_message__(self):
        self.logger.debug("sr_log2source __on_message__")

        # is the log message for this cluster

        if not 'from_cluster' in self.msg.headers or self.msg.headers['from_cluster'] != self.cluster :
           self.logger.debug("skipped : not for cluster %s" % self.cluster)
           self.logger.debug("hdr from_cluster %s" % self.msg.headers['from_cluster'])
           return False

        # is the log message from a source on this cluster

        if not 'source' in self.msg.headers or not self.msg.headers['source'] in self.source_users:
           self.logger.debug("skipped : source not in %s" % self.source_users)
           return False

        # yup this is one message we want to ship to our source

        # user provided an on_message script

        ok = True

        if self.on_message : ok = self.on_message(self)

        return ok

    # =============
    # __on_post__ internal posting
    # =============

    def __on_post__(self):

        # invoke on_post when provided

        if self.on_post :
           ok = self.on_post(self)
           if not ok: return ok

        # should always be ok

        ok = self.msg.publish( )

        return ok

    def overwrite_defaults(self):

        # overwrite defaults

        self.broker               = self.manager
        self.exchange             = 'xreport'
        self.topic_prefix         = 'v02.report'
        self.subtopic             = '#'

    # =============
    # process message  
    # =============

    def process_message(self):

        try  :
                 ok, self.msg = self.consumer.consume()
                 if not ok : return ok

                 self.logger.debug("Received topic   %s" % self.msg.topic)
                 self.logger.debug("Received notice  %s" % self.msg.notice)
                 self.logger.debug("Received headers %s" % self.msg.hdrstr)

                 # invoke __on_message__

                 ok = self.__on_message__()
                 if not ok : return ok

                 # ok ship it back to the user exchange 

                 self.msg.exchange = 'xr_' + self.msg.headers['source']

                 # invoke __on_post__

                 ok = self.__on_post__()

                 return ok

        except :
                 (stype, svalue, tb) = sys.exc_info()
                 self.logger.error("Type: %s, Value: %s,  ..." % (stype, svalue))
                 return False

        return ok


    def run(self):

        # present basic config

        self.logger.info("sr_log2source run")
        self.logger.info("AMQP  broker(%s) user(%s) vhost(%s)" % \
                        (self.broker.hostname,self.broker.username,self.broker.path) )

        for tup in self.bindings:
            e,k =  tup
            self.logger.info("AMQP  input :    exchange(%s) topic(%s)" % (e,k) )

        self.logger.info("\nsource users = %s" % self.source_users)


        # connect and loop/process messages

        self.connect()

        while True :
              ok = self.process_message()


    def reload(self):
        self.logger.info("%s reload" % self.program_name)
        self.close()
        self.configure()
        self.run()

    def start(self):
        self.logger.info("%s start" % self.program_name)
        self.run()

    def stop(self):
        self.logger.info("%s stop" % self.program_name)
        self.close()
        os._exit(0)
                 

# ===================================
# self test
# ===================================

class test_logger:
      def silence(self,str):
          pass
      def __init__(self):
          self.debug   = print
          self.error   = print
          self.info    = print
          self.warning = print

def test_sr_log2source():

    logger = test_logger()

    yyyy   = time.strftime("%Y",time.gmtime())
    opt1   = 'accept .*' + yyyy + '.*'
    opt2   = 'reject .*'
    opt3   = 'on_message ./on_msg_test.py'
    opt4   = 'on_post ./on_pst_test.py'

    # here an example that calls the default_on_message...
    # then process it if needed
    f      = open("./on_msg_test.py","w")
    f.write("class Transformer(object): \n")
    f.write("      def perform(self, parent ):\n")
    f.write("          parent.msg.mtypej = 'transformed'\n")
    f.write("          return True\n")
    f.write("transformer = Transformer()\n")
    f.write("self.on_message = transformer.perform\n")
    f.close()

    f      = open("./on_pst_test.py","w")
    f.write("class Transformer(object): \n")
    f.write("      def perform(self, parent ):\n")
    f.write("          parent.msg.mtypek = 'transformed'\n")
    f.write("          return True\n")
    f.write("transformer = Transformer()\n")
    f.write("self.on_post = transformer.perform\n")
    f.close()

    # setup sr_log2source to catch repost N log messages

    N = 10

    log2source         = sr_log2source()
    log2source.logger  = logger
    log2source.debug   = True

    log2source.user_cache_dir = os.getcwd()

    # because following options have accept/reject
    log2source.masks   = []
    log2source.option( opt1.split()  )
    log2source.option( opt2.split()  )

    log2source.option( opt3.split()  )
    log2source.option( opt4.split()  )

    # ==================
    # define YOUR BROKER HERE

    ok, details = log2source.credentials.get("amqp://tfeed@localhost/")
    if not ok :
       print("UNABLE TO PERFORM TEST")
       print("Need feeder privileges to a broker")
       sys.exit(1)
    log2source.broker = details.url

    # ==================
    # define a source_users list here

    log2source.source_users = ['tsource']

    # ==================
    # define the matching cluster here
    log2source.cluster = 'DDI.CMC'

    log2source.connect()

    # process N messages

    i = 0
    j = 0
    k = 0
    while True :
          if log2source.process_message():
             if log2source.msg.mtypej == 'transformed': j += 1
             if log2source.msg.mtypek == 'transformed': k += 1
             i = i + 1
          if i == N: break

    os.unlink(log2source.consumer.queuepath)

    log2source.close()

    if j != N or k != N :
       print("sr_log2source TEST Failed 1")
       sys.exit(1)

    print("sr_log2source TEST PASSED")

    os.unlink('./on_msg_test.py')
    os.unlink('./on_pst_test.py')

    sys.exit(0)

# ===================================
# MAIN
# ===================================

def main():

    action = None
    args   = None
    config = None

    if len(sys.argv) > 1 :
       action = sys.argv[-1]
       args   = sys.argv[1:-1]

    if len(sys.argv) > 2 : 
       config    = sys.argv[-2]
       cfg       = sr_config()
       cfg.defaults()
       cfg.general()
       ok,config = cfg.config_path('log2source',config,mandatory=False)
       if ok     : args = sys.argv[1:-2]
       if not ok : config = None

    log2source = sr_log2source(config,args)

    if action != 'TEST' and  not log2source.log_daemons :
       log2source.logger.info("sr_log2source will not run (log_daemons), action '%s' ignored " % action)
       sys.exit(0)

    if   action == 'reload' : log2source.reload_parent()
    elif action == 'restart': log2source.restart_parent()
    elif action == 'start'  : log2source.start_parent()
    elif action == 'stop'   : log2source.stop_parent()
    elif action == 'status' : log2source.status_parent()
    elif action == 'TEST'   : test_sr_log2source()
    else :
           log2source.logger.error("action unknown %s" % action)
           sys.exit(1)

    sys.exit(0)

# =========================================
# direct invocation
# =========================================

if __name__=="__main__":
   main()
